src/main/kotlin/com/openlattice/graph/partioning/RepartitioningJob.kt

```diff
@@ -0,0 +1,352 @@
+    constructor(
+            id: UUID?,
+            taskId: Long?,
+            status: JobStatus,
+            progress: Byte,
+            hasWorkRemaining: Boolean,
+            result: Long?,
+            state: RepartitioningJobState
+    ) : this(state) {
+        initialize(id, taskId, status, progress, hasWorkRemaining, result)
+    }
+
+    constructor(
+            entitySetId: UUID,
+            oldPartitions: List<Int>,
+            newPartitions: Set<Int>
+    ) : this(RepartitioningJobState(entitySetId, oldPartitions, newPartitions))
+
+    private var phase = 0
+
+    @Transient
+    private lateinit var hds: HikariDataSource
+
+    @Transient
+    private lateinit var entitySets: IMap<UUID, EntitySet>
+
+    private val currentlyMigratingPartition: Int
+        get() = state.oldPartitions[state.currentlyMigratingPartitionIndex]
+
+    @JsonIgnore
+    override fun setHazelcastInstance(hazelcastInstance: HazelcastInstance) {
+        super.setHazelcastInstance(hazelcastInstance)
+        this.entitySets = HazelcastMap.ENTITY_SETS.getMap(hazelcastInstance)
+    }
+
+    @JsonIgnore
+    override fun setHikariDataSource(hds: HikariDataSource) {
+        this.hds = hds
+    }
+
+    override fun initialize() {
+        state.needsMigrationCount = getNeedsMigrationCount()
+    }
+
+    override fun processNextBatch() {
+        if (state.needsMigrationCount == 0L) {
+            hasWorkRemaining = false
+            return
+        }
+
+        /**
+         * Do an INSERT INTO ... SELECT FROM to re-partition the data.
+         *
+         * Entity key id depenedent operations will not see data, until data has been inserted to the appropriate partition.
+         */
+
+        state.repartitionCount += repartition(REPARTITION_DATA_SQL)
+        state.repartitionCount += repartition(REPARTITION_IDS_SQL)
+        state.repartitionCount += repartition(REPARTITION_EDGES_SQL)
+
+        /**
+         * Phase 1
+         * Delete data whose partition doesn't match it's computed partition.
+         */
+        if (phase == 1) {
+            state.deleteCount += delete(DELETE_DATA_SQL)
+            state.deleteCount += delete(DELETE_IDS_SQL)
+            state.deleteCount += delete(DELETE_EDGES_SQL)
+        }
+
+        result = state.repartitionCount + state.deleteCount
+        hasWorkRemaining = (++state.currentlyMigratingPartitionIndex < state.oldPartitions.size)
+
+        //TODO: Consider adding completion hook to distributable jobs framework
+        //Once we are done, set the partitions.
+        if (!hasWorkRemaining && phase == 0) {
+            setPartitions(state.entitySetId, state.newPartitions)
+            //The 4*getCount was an estimate, we remove estimate and addin updated value.
+            state.needsMigrationCount /= 2
+            state.needsMigrationCount += getNeedsMigrationCount()
+            state.currentlyMigratingPartitionIndex = 0
+            hasWorkRemaining = true
+            phase = 1
```

Comment:
^ okay, maybe phase 0 doesn't mean complete...

Reply:
Phase 0 populates data into new partitions, we then mark new partitions, then phase 1 merges in any changes since phase 0.